### PR TITLE
Add LocalizationSetContentPickerField

### PR DESCRIPTION
**Pull request by jptissot**
Fixes: #3821 

This PR adds a field similar to the ContentPicker field that allows you to target a LocalizationSet instead of a ContentItemId.

This allows us to support a hierarchy of content where the parent content might be localized and the child content wants to target the LocalizationSet to be able to get the parent dynamically based on culture.

What has been done:
- Refactor the ContentPicker's javascript code to support both the ContentPicker and LocalizationSetContentPicker
- Add LocalizationSetContentPickerField only when ContentLocalization feature is enabled
- Add liquid filter that gets the list of ContentItems that matches the set(s) for the current locale.
- Create a LocalizationSetContentPicker search controller that return a single ContentItem per LocalizationSet
- Debounce the javascript ContentPicker search function so less requests are sent to the server
- Trigger an event on body when the ContentPicker or LocalizedContentPicker is initialized so other unlerated parts of the code can hook into it (@Skrypt I think this was one of your use cases)
- Add documentation

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Controllers/LocalizedContentPickerAdminController.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            var partFieldDefinition = _contentDefinitionManager.GetPartDefinition(part)?.Fields
+                .FirstOrDefault(f => f.Name == field);
+
+            var fieldSettings = partFieldDefinition?.Settings.ToObject<LocalizedContentPickerFieldSettings>();
+            if (fieldSettings == null)
+            {
+                return BadRequest("Unable to find field definition");
+            }
+
+            var dbQuery = _session.Query<ContentItem, ContentItemIndex>()
```
Comment (jptissot) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Controllers/LocalizedContentPickerAdminController.cs]: Should additional Authorization checks be performed here ?

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Controllers/ContentPickerAdminController.cs [hunk 2]
```diff
@@ -23,7 +26,8 @@ IEnumerable<IContentPickerResultProvider> resultProviders
             _resultProviders = resultProviders;
         }
 
-        public async Task<IActionResult> List(string part, string field, string query)
+        [HttpGet]
+        public async Task<IActionResult> SearchContentItems(string part, string field, string query)
         {
```
Comment (jptissot) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Controllers/ContentPickerAdminController.cs]: Should this method do additional Authorization checks ?

**Comment by jean** [comment]
What will happen if translations come or go? E.g. I'm in Spanish, when I pick an item I'd like the Spanish translation to show _if available_, falling back to English if the translation doesn't exist, or if the translation has been unpublished. I wouldn't like to pick a translation per language. Could you confirm whether this is how the current implementation works?

**Comment by jptissot** [comment]
Hi @jean,

> What will happen if translations come or go? E.g. I'm in Spanish, when I pick an item I'd like the Spanish translation to show _if available_, falling back to English if the translation doesn't exist, or if the translation has been unpublished. I wouldn't like to pick a translation per language. Could you confirm whether this is how the current implementation works?

The `LocalizedContentPicker` is storing the `LocalizationSet` of the `ContentItem` it references. A `LocalizationSet` is a shared identifier that is assigned to a `ContentItem` that has a `LocalizationPart`.  So your Spanish version of content and English version of content share a `LocalizationSet`. 

This field makes no assumptions on which `ContentItem` is displayed in the frontend. It will give you the LocalizationSet and you have to get the items you want to display in the frontend. 

I already added a [LiquidFilter](https://github.com/OrchardCMS/OrchardCore/pull/3819/files#diff-3848b9445eef84149a46e0b9f5fa8499R31) that will get the ContentItem of the LocalizationSet based on the current thread culture. We will probably want to add other fitlers to support the use case you mentionned (falling back to a default locale for example).

Hope this helps :)

**Comment by jptissot** [comment]
@sebastienros I thought about this and it could potentially be possible to do this with the ContentPicker. The frontend would have a new filter / helper to get the localized version of a specific ContentItem. It would be up to the frontend to know.

**Comment by jean** [comment]
> We will probably want to add other [filters] to support the use case you [mentioned] (falling back to a default locale for example).

I think that this should be transparent in the default case. I.e. items with and without translations are referenced identically, and if available the translation for the configured culture is rendered. 

If necessary, it remains possible to reference specific translation(s), but it should be possible to add translations to a monolingual site without rewriting templates.

**Comment by jptissot** [comment]
I found a couple of bugs: 

- Missing await calls in service
- Ordering Localization sets does not work

**Comment by jptissot** [comment]
Fixed a couple of bugs but Ordering still does not work. I will have to investigate.

**Comment by jptissot** [comment]
@Skrypt how does ordering work with the ContentPicker ?

**Comment by Skrypt** [comment]
I think it reorders naturally the JSON elements simply. But I'm guessing I've not looked.

**Comment by jptissot** [comment]
@sebastienros should this be named CultureAgnosticContentPicker instead ?

**Comment by jptissot** [comment]
@sebastienros this is ready for a review.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Views/LocalizationSetContentPickerField.Edit.cshtml [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+<script asp-src="https://cdnjs.cloudflare.com/ajax/libs/Vue.Draggable/2.14.1/vuedraggable.min.js" debug-src="https://cdnjs.cloudflare.com/ajax/libs/Vue.Draggable/2.14.1/vuedraggable.js" asp-name="vuedraggable" depends-on="vuejs, sortable" at="Foot"></script>
+<script asp-src="~/OrchardCore.ContentFields/Scripts/vue-multiselect-wrapper.min.js" debug-src="~/OrchardCore.ContentFields/Scripts/vue-multiselect-wrapper.js" asp-name="vue-multiselect-wrapper" at="Foot" depends-on="vuejs, vuemultiselect, sortable"></script>
+<style asp-src="~/OrchardCore.ContentFields/Styles/vue-multiselect-wrapper.min.css" debug-src="~/OrchardCore.ContentFields/Styles/vue-multiselect-wrapper.css"></style>
+
+<label asp-for="LocalizationSets">@Model.PartFieldDefinition.DisplayName()</label>
+
+<div id="@vueElementId" class="vue-multiselect"  data-editor-type="LocalizationSetContentPicker" data-selected-items="@selectedItems" data-tenant-path="@tenantPath" data-search-url="ContentFields/SearchLocalizationSets?part=@partName&field=@fieldName" data-multiple="@multiple">
```
Comment (jptissot) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Views/LocalizationSetContentPickerField.Edit.cshtml]: I personally like this pattern to add a class and data-attributes to elements and use the  document.querySelectorAll() to find elements and instanciate the plugin. Is this a good pattern ? It removes the JS call at the bottom of the file and it's easier to wrap it in a Self-Executing Anonymous Function.

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/DefaultContentLocalizationManager.cs [hunk 4]
```diff
@@ -100,5 +108,77 @@ public async Task<ContentItem> LocalizeAsync(ContentItem content, string targetC
             _session.Save(cloned);
             return cloned;
         }
+
+        public async Task<IDictionary<string, ContentItem>> DeduplicateContentItemsAsync(IEnumerable<ContentItem> contentItems)
```
Comment (jptissot) [inline src/OrchardCore.Modules/OrchardCore.ContentLocalization/DefaultContentLocalizationManager.cs]: Maybe I misunderstood the naming convention of Async. 

Should a method name end with Async when they don't await the last statement ? @jtkech

Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.ContentLocalization/DefaultContentLocalizationManager.cs]: It's just a different way to write your method body, as soon as your method is seen by the caller as an awaitable `Task` the convention is applied.

Yes, if in all your code paths you don't need to await others async methods (as in a loop or to init a var to be used later), you don't have to use the `async` keyword and all your return paths can directly return a `Task` (e.g another async method, a `Task.FromResult` in another place ...) without awaiting it (preventing some more allocations).

But as said it's just a different way, if the last statement is awaited the convention is still applied.

**Pull request merged**